docs/contributing/using-instrumenter-api.md

```diff
@@ -1 +1,468 @@
+convenient `getRelatedOperations()` method to find all spans that should be linked to the newly
+created one. In reality, you will most likely have to construct `SpanContext` by extracting
+information from a data structure representing request headers or metadata.
+
+You can add a `SpanLinksExtractor` to the `InstrumenterBuilder` by using
+the `addSpanLinksExtractor()` method.
+
+### Discard wrapper exception types with the `ErrorCauseExtractor`
+
+Sometimes when an error occurs, the real cause is hidden behind several "wrapper" exception types -
+for example, an `ExecutionException` or a `CompletionException` from the Java standard library. By
+default, the known wrapper exception types from the JDK are removed from the captured error. To
+remove other wrapper exceptions, like the ones provided by the instrumented library, you can
+implement the `ErrorCauseExtractor`. It has only one method `extractCause()` that is responsible for
+stripping the unnecessary exception layers and extracting the actual error that caused the
+processing to fail. It accepts a `Throwable` and returns a `Throwable`.
+
+Consider the following example:
+
+```java
+class MyErrorCauseExtractor implements ErrorCauseExtractor {
+
+  @Override
+  public Throwable extractCause(Throwable error) {
+    if (error instanceof MyLibWrapperException && error.getCause() != null) {
+      error = error.getCause();
+    }
+    return ErrorCauseExtractor.jdk().extractCause(error);
+  }
+}
+```
+
+The example `ErrorCauseExtractor` implementation checks whether the error is an instance
+of `MyLibWrapperException` and has a cause - if so, it unwraps it. The `error.getCause() != null`
+check is rather important here: if the extractor did not verify this it could accidentally remove
+the whole exception, making the instrumentation miss an error and thus radically changing the
+captured telemetry. Next, the extractor falls back to the default `jdk()` implementation that
+removes the known JDK wrapper exception types.
+
+You can set the `ErrorCauseExtractor` in the `InstrumenterBuilder` using
+the `setErrorCauseExtractor()` method.
+
+### Provide custom operation start and end times using the `StartTimeExtractor` and `EndTimeExtractor`
+
+In some cases, the instrumented library provides a way to retrieve accurate timestamps of when the
+operation starts and ends. The `StartTimeExtractor` and `EndTimeExtractor` interfaces can be used to
+feed this information into OpenTelemetry trace and metrics data. Ether both time extractors or none
+at all need to be provided to the `InstrumenterBuilder`. It is crucial to avoid a situation where
+one time measurement uses the library timestamp and the other the internal OpenTelemetry SDK clock,
+as it would result in inaccurate telemetry.
+
+The `StartTimeExtractor` can only extract the timestamp from the request. The `EndTimeExtractor`
+accepts the request, an optional response, and an optional `Throwable` error. Consider the following
```

Comment:
Question: why do we need throwable here?

Reply:
It's just for symmetry -- all places that accept a `RESPONSE` accept a `Throwable` error too.